Diff Krun results files

snim2 opened the pull request:
This PR adds a script which allows the user to diff two Krun results files, based on the algorithm that generated Figures 9 / 11 in the warmup paper.

The output is a LaTeX table, similar to the one produced by `table_classification_summaries_others`, which shows the results of the *second* input file, and some extra symbols showing which parts of the table have improved / worsened, etc. 

The symbol next to the benchmark name shows the 'intersection' of the other measures for that benchmark. Where NSS's appear in the summaries, we sometimes cannot compare before/after results, or sometimes we can only say that the two results differ.

The script dumps a human-readable JSON file with the results of the summary and the diff. This JSON file can then be read in by the script to produce a LaTeX file (rather than recomputing the diff and summaries directly from the Krun results files). This is probably less useful for end-users and more useful for debugging the script.

Also, the fix for Issue #15 turned out to have a bug in tables which only took up a single page. This PR contains a fix for that bug.

### CLI options

```
$ ./bin/diff_results --help
usage: diff_results [-h] [-o OUTPUT] [-j JSON] [--num-splits NUM_SPLITS]
                    [--without-preamble]
                    (-s SUMMARY | -r RESULTS_FILES RESULTS_FILES)

Diff two Krun results files. Input files to this script should already have outliers and changepoints marked (i.e. the mark_outliers_in_json and mark_changepoints_in_json scripts should already have been run).

Example usage:

	$ python diff_results -f before.json.bz2 after.json.bz2 -o diff.tex
	$ python diff_results -s diff_summary.json -o diff.tex

optional arguments:
  -h, --help            show this help message and exit
  -o OUTPUT, --output OUTPUT
                        LaTeX file in which to write diff summary.
  -j JSON, --json JSON  JSON file in which to write diff summary.
  --num-splits NUM_SPLITS, -n NUM_SPLITS
                        Number of horizontal splits.
  --without-preamble    Write out only the table (for inclusion in a separate document).
  -s SUMMARY, --summary SUMMARY
                        Read summary data from JSON file rather than generating from a two original results files.
  -r RESULTS_FILES RESULTS_FILES, --results-files RESULTS_FILES RESULTS_FILES
                        Exactly two Krun result files (with outliers and changepoints).
```

### Examples

**Before input data**: [before_table.pdf](https://github.com/softdevteam/warmup_stats/files/1644919/before_table.pdf)

**After1 input data**: [after_table.pdf](https://github.com/softdevteam/warmup_stats/files/1644920/after_table.pdf)

**After2 input data**: [after2_table.pdf](https://github.com/softdevteam/warmup_stats/files/1644922/after2_table.pdf)

**Diff Before -> After1**: [diff.pdf](https://github.com/softdevteam/warmup_stats/files/1644927/diff.pdf)

**Diff Before -> After2**: [diff2.pdf](https://github.com/softdevteam/warmup_stats/files/1644928/diff2.pdf)

Fixes #7

fsfod commented:
It would be helpful to somehow show how much it changed in some way.

ltratt commented:
At the moment, the key says $\uparrow$ means "improved", but it's actually used for things that have gotten slower!

ltratt commented:
Re: #3. I think we could use something like uparrow, downarrow, tilde, and blank for the 4 sub-classifications I mentioned.

snim2 commented:
Re 1. I'm with you now. Re 3. I'm not sure, "some things got better and other things got worse" isn't really the same as "nothing changed". Let me give this some more thought.

ltratt commented:
Agreed -- that's why I had a "no change" fourth category.

snim2 commented:
Ugh, sorry, I missed the semi-colon there!

snim2 commented:
This isn't quite finished yet, but it's probably worth getting some early feedback, this is the first of the diffs above: [progress.pdf](https://github.com/softdevteam/warmup_stats/files/1647230/progress.pdf)

Cell backgrounds are in, and the benchmark name has the "overall" measure, as specified above. Green means "improved", red "worsened", white (blank) means "no change, or couldn't compare due to NSS", yellow means "some different some not". `jsonlua_decode` seems dubious, so I'm investigating that before pushing the latest changes.

We are now showing whether classifications are better/worse, whereas before we could only say they were same/different.

I haven't start the "show how much better / worse the measure improved", but I think that should be straight forward.  

There is also an issue with the coloured cells being a bit too close to their neighbours, but I'm reaching the limit of my LaTeX skills with that, changing the `tabcolsep` length doesn't seem to help.

Anything else?

ltratt commented:
This looks like a decent start. I have a major and minor comment.

The major comment is that I don't think the way we're using colours totally makes sense. e.g. what does the yellow of jsonlua_decode steady perf mean? I dunno. I think that red/green/white make sense in all fields, but yellow only makes sense on the benchmark name. Put another way, if a benchmark name is yellow, it means that at least one other box on that row should be red and at least one other box on that row should be green.

As a minor comment, I think we need to choose less bright colours or (probably better?) use translucency though -- at the moment the backgrounds drown out the text and, in particular, the histograms. I'm not too worried about the size of the cell colouring as that doesn't affect comprehensibility.

snim2 commented:
Changes since the last examples:
  * Colours are improved, so the text is more readable
  * Spacing between table cells is fixed
  * Where a measure is better / worse, underneath the confidence interval or IQR is a line of text that says `$\delta=...$` and states the absolute difference between the median/mean of the before/after results.
  * The faster/slower bug is fixed.
  * Some slightly tidier code.

One thing I have not changed is using yellow in the steady iter / time / perf measures. Yellow here means "we could not simply compare the before/after for this measure, because one of the data sets contained an NSS". E.g. for `jsonlua_decode`, the "before" data has 1 NSS, 2 warmup and 7 flat, whereas the "after" data has 30 flat (nevertheless, the confidence intervals for the classifications overlap).

New diff before->after1: [diff_new.pdf](https://github.com/softdevteam/warmup_stats/files/1657060/diff_new.pdf)

ltratt commented:
This looks nice, and I agree with the "yellow" comment.

My only comment is a minor one: should we move the $\delta$ to *before* the confidence interval? When I first looked at things, I thought it applied to the confidence interval.

Oh, and I'm not sure the arrows are doing anything useful any more? I think we could get rid of them entirely, and rely on the colours?

snim2 commented:
OK, how about this: [diff_new2.pdf](https://github.com/softdevteam/warmup_stats/files/1657556/diff_new2.pdf)

ltratt commented:
All I think we need now is a legend explaining the colours and we're done! Could be as simple as "Improved" (with a red background) and so on.

snim2 commented:
Ah yes, good point...

snim2 commented:
With legend: [diff_new_with_legend.pdf](https://github.com/softdevteam/warmup_stats/files/1657602/diff_new_with_legend.pdf)

ltratt commented:
Works for me!

If you squash, I can merge.

snim2 commented:
Squashed!
